File: app/src/main/java/de/grobox/transportr/map/MapViewModel.java
package de.grobox.transportr.map;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.util.Log;
import android.widget.Toast;

import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.geometry.LatLngBounds;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.inject.Inject;
import javax.inject.Singleton;

import de.grobox.transportr.R;
import de.grobox.transportr.TransportrApplication;
import de.grobox.transportr.data.locations.FavoriteLocation;
import de.grobox.transportr.data.locations.LocationRepository;
import de.grobox.transportr.data.searches.SearchesRepository;
import de.grobox.transportr.favorites.trips.SavedSearchesViewModel;
import de.grobox.transportr.locations.WrapLocation;
import de.grobox.transportr.networks.TransportNetworkManager;
import de.grobox.transportr.utils.SingleLiveEvent;

@Singleton
public class MapViewModel extends SavedSearchesViewModel {

	private final GpsController gpsController;

	private final MutableLiveData<Integer> peekHeight = new MutableLiveData<>();
	private final MutableLiveData<LatLng> selectedLocationClicked = new MutableLiveData<>();
	private final MutableLiveData<LatLngBounds> updatedLiveBounds = new MutableLiveData<>();
	private final MutableLiveData<WrapLocation> selectedLocation = new MutableLiveData<>();
	private final SingleLiveEvent<WrapLocation> findNearbyStations = new SingleLiveEvent<>();
	private final SingleLiveEvent<Boolean> nearbyStationsFound = new SingleLiveEvent<>();

	final SingleLiveEvent<Void> mapClicked = new SingleLiveEvent<>();
	final SingleLiveEvent<Void> markerClicked = new SingleLiveEvent<>();
	final LiveData<LatLngBounds> liveBounds = Transformations.switchMap(getLocations(), this::switchMap);

	@Inject
	MapViewModel(TransportrApplication application, TransportNetworkManager transportNetworkManager, LocationRepository locationRepository,
	             SearchesRepository searchesRepository, GpsController gpsController) {
		super(application, transportNetworkManager, locationRepository, searchesRepository);
		this.gpsController = gpsController;
	}

	GpsController getGpsController() {
		return gpsController;
	}

	LiveData<Integer> getPeekHeight() {
		return peekHeight;
	}

	public void setPeekHeight(int peekHeight) {
		this.peekHeight.setValue(peekHeight);
	}

	LiveData<LatLng> getSelectedLocationClicked() {
		return selectedLocationClicked;
	}

	public void selectedLocationClicked(LatLng latLng) {
		selectedLocationClicked.setValue(latLng);
	}

	void selectLocation(@Nullable WrapLocation location) {
		selectedLocation.setValue(location);
	}

	LiveData<WrapLocation> getSelectedLocation() {
		return selectedLocation;
	}

	public void findNearbyStations(WrapLocation location) {
		findNearbyStations.setValue(location);
	}

	LiveData<WrapLocation> getFindNearbyStations() {
		return findNearbyStations;
	}

	void setNearbyStationsFound(boolean found) {
		nearbyStationsFound.setValue(found);
	}

	public LiveData<Boolean> nearbyStationsFound() {
		return nearbyStationsFound;
	}

	private MutableLiveData<LatLngBounds> switchMap(List<FavoriteLocation> input) {
		if (input == null) {
			updatedLiveBounds.setValue(null);
		} else {
			Set<LatLng> points = new HashSet<>();
			for (FavoriteLocation location : input) {
				if (location.hasLocation()) points.add(location.getLatLng());
			}
			WrapLocation gpsLocation = gpsController.getWrapLocation();
			if (gpsLocation != null && gpsLocation.hasLocation()) points.add(gpsLocation.getLatLng());
			if (points.size() < 2) {
				updatedLiveBounds.setValue(null);
			} else {
				updatedLiveBounds.setValue(new LatLngBounds.Builder().includes(new ArrayList<>(points)).build());
			}
		}
		return updatedLiveBounds;
	}

	void setGeoUri(Uri geoUri) {
		WrapLocation location = getWrapLocation(geoUri.toString());
		if (location != null) {
			selectLocation(location);
		} else {
			Log.w(MapViewModel.class.getSimpleName(), "Invalid geo intent: " + geoUri.toString());
			Toast.makeText(getApplication().getApplicationContext(), R.string.error_geo_intent, Toast.LENGTH_SHORT).show();
		}
	}

	@Nullable
	@VisibleForTesting
	WrapLocation getWrapLocation(String geoUri) {
		Pattern pattern = Pattern.compile("^geo:(-?\\d{1,3}(\\.\\d{1,8})?),(-?\\d{1,3}(\\.\\d{1,8})?).*");
		Matcher matcher = pattern.matcher(geoUri);
		if (matcher.matches()) {
			double lat = Double.valueOf(matcher.group(1));
			double lon = Double.valueOf(matcher.group(3));
			if (lat == 0 && lon == 0) return null;
			return new WrapLocation(new LatLng(lat, lon));
		}
		return null;
	}

}

File: app/src/main/java/de/grobox/transportr/map/MapViewModel.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.map

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import android.net.Uri
import android.util.Log
import android.widget.Toast
import com.mapbox.mapboxsdk.geometry.LatLng
import com.mapbox.mapboxsdk.geometry.LatLngBounds
import de.grobox.transportr.R
import de.grobox.transportr.TransportrApplication
import de.grobox.transportr.data.locations.FavoriteLocation
import de.grobox.transportr.data.locations.LocationRepository
import de.grobox.transportr.data.searches.SearchesRepository
import de.grobox.transportr.favorites.trips.SavedSearchesViewModel
import de.grobox.transportr.locations.WrapLocation
import de.grobox.transportr.networks.TransportNetworkManager
import de.grobox.transportr.utils.IntentUtils
import de.grobox.transportr.utils.SingleLiveEvent
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
internal class MapViewModel @Inject internal constructor(
        application: TransportrApplication,
        transportNetworkManager: TransportNetworkManager,
        locationRepository: LocationRepository,
        searchesRepository: SearchesRepository,
        val gpsController: GpsController) : SavedSearchesViewModel(application, transportNetworkManager, locationRepository, searchesRepository) {

    private val peekHeight = MutableLiveData<Int>()
    private val selectedLocationClicked = MutableLiveData<LatLng>()
    private val updatedLiveBounds = MutableLiveData<LatLngBounds>()
    private val selectedLocation = MutableLiveData<WrapLocation>()
    private val findNearbyStations = SingleLiveEvent<WrapLocation>()
    private val nearbyStationsFound = SingleLiveEvent<Boolean>()

    val mapClicked = SingleLiveEvent<Void>()
    val markerClicked = SingleLiveEvent<Void>()
    val liveBounds: LiveData<LatLngBounds> = Transformations.switchMap<List<FavoriteLocation>, LatLngBounds>(locations, this::switchMap)
    var transportNetworkWasChanged = false

    fun getPeekHeight(): LiveData<Int> {
        return peekHeight
    }

    fun setPeekHeight(peekHeight: Int) {
        this.peekHeight.value = peekHeight
    }

    fun getSelectedLocationClicked(): LiveData<LatLng> {
        return selectedLocationClicked
    }

    fun selectedLocationClicked(latLng: LatLng) {
        selectedLocationClicked.value = latLng
        // reset the selected location right away, observers will ignore this update
        selectedLocationClicked.value = null
    }

    fun selectLocation(location: WrapLocation?) {
        selectedLocation.value = location
        // do not reset the selected location right away, will break incoming geo intent
        // the observing fragment will call clearSelectedLocation() instead when it is done
        gpsController.updateGpsState(isTracking = false)
    }

    fun clearSelectedLocation() {
        selectedLocation.postValue(null)
    }

    fun getSelectedLocation(): LiveData<WrapLocation> {
        return selectedLocation
    }

    fun findNearbyStations(location: WrapLocation) {
        findNearbyStations.value = location
    }

    fun getFindNearbyStations(): LiveData<WrapLocation> {
        return findNearbyStations
    }

    fun setNearbyStationsFound(found: Boolean) {
        nearbyStationsFound.value = found
    }

    fun nearbyStationsFound(): LiveData<Boolean> {
        return nearbyStationsFound
    }

    fun setGeoUri(geoUri: Uri) {
        val location = IntentUtils.getWrapLocation(geoUri.toString())
        if (location != null) {
            selectLocation(location)
        } else {
            Log.w(MapViewModel::class.java.simpleName, "Invalid geo intent: " + geoUri.toString())
            Toast.makeText(getApplication<Application>().applicationContext, R.string.error_geo_intent, Toast.LENGTH_SHORT).show()
        }
    }

    private fun switchMap(input: List<FavoriteLocation>?): MutableLiveData<LatLngBounds> {
        if (input == null) {
            updatedLiveBounds.setValue(null)
        } else {
            val points = input
                    .filter { it.hasLocation() }
                    .map { it.latLng as LatLng }
                    .toMutableSet()
            home.value?.let { if (it.hasLocation()) points.add(it.latLng) }
            work.value?.let { if (it.hasLocation()) points.add(it.latLng) }
            gpsController.getWrapLocation()?.let { if (it.hasLocation()) points.add(it.latLng) }
            if (points.size < 2) {
                updatedLiveBounds.setValue(null)
            } else {
                updatedLiveBounds.setValue(LatLngBounds.Builder().includes(ArrayList(points)).build())
            }
        }
        return updatedLiveBounds
    }

}
